compiler/rustc_builtin_macros/src/deriving/smart_ptr.rs

```diff
@@ -143,31 +152,175 @@ pub fn expand_deriving_smart_ptr(
 
     // Find the `#[pointee]` parameter and add an `Unsize<__S>` bound to it.
     let mut impl_generics = generics.clone();
+    let pointee_ty_ident = generics.params[pointee_param_idx].ident;
+    let mut self_bounds;
     {
         let p = &mut impl_generics.params[pointee_param_idx];
+        self_bounds = p.bounds.clone();
         let arg = GenericArg::Type(s_ty.clone());
         let unsize = cx.path_all(span, true, path!(span, core::marker::Unsize), vec![arg]);
         p.bounds.push(cx.trait_bound(unsize, false));
         let mut attrs = thin_vec![];
         swap(&mut p.attrs, &mut attrs);
         p.attrs = attrs.into_iter().filter(|attr| !attr.has_name(sym::pointee)).collect();
     }
+    // We should not set default values to constant generic parameters
+    // and write out bounds that indirectly involves `#[pointee]`.
+    for (idx, (params, orig_params)) in
+        impl_generics.params.iter_mut().zip(&generics.params).enumerate()
+    {
+        if idx == pointee_param_idx {
+            continue;
+        }
```

Comment:
Why can't this be consolidated with the `#[pointee]` param arm? Iterating over all non-pointee params and then separately handling all pointee params is really difficult to understand, and it allows for subtle discrepancies in behavior IMO.

Reply:
If this needs to be separate, there should be a detailed explanation about how bounds on non-`#[pointee]` params and `#[pointee]` params are propagated differently.

Reply:
I added a few comments to explain the disparity, with illustration.